src/opencl_keystore_fmt_plug.c

```diff
@@ -0,0 +1,587 @@
+	{NULL}
+};
+
+// these to pass to kernel
+typedef struct {
+	uint32_t length;
+	uint8_t  pass[PLAINTEXT_LENGTH*2];
+} keystore_password;
+
+typedef struct {
+	uint32_t key[BINARY_SIZE/4];
+} keystore_hash;
+
+typedef struct {
+	uint32_t length;
+	uint8_t  salt[SALT_LENGTH];
+} keystore_salt;
+
+// this for use here
+static struct custom_salt {
+	int length;
+	unsigned char salt[SALT_LENGTH];
+} *cur_salt;
+
+static struct fmt_main   *self;
+
+static size_t insize,
+       	   	  outsize,
+			  saltsize;
+
+static keystore_password *inbuffer;
+static keystore_hash     *outbuffer;
+static keystore_salt      saltbuffer;
+static cl_mem mem_in,
+              mem_out,
+			  mem_salt;
+
+static cl_int cl_err;
+
+// This file contains auto-tuning routine(s). Has to be included after formats definitions.
+#include "opencl-autotune.h"
+#include "memdbg.h"
+
+static const char * warn[] = {
+	"xfer: ",  ", crypt: ",  ", xfer: "
+};
+
+/* ------- Helper functions ------- */
+static size_t get_task_max_work_group_size()
+{
+	return autotune_get_task_max_work_group_size(FALSE, 0, crypt_kernel);
+}
+
+static void create_clobj(size_t gws, struct fmt_main *self)
+{
+	insize = sizeof(keystore_password) * gws;
+	outsize = sizeof(keystore_hash) * gws;
+	saltsize = sizeof(keystore_salt);
+
+	inbuffer  = mem_calloc(1, insize);
+	outbuffer = mem_alloc(outsize);
+
+	/// Allocate memory
+	mem_in =
+	    clCreateBuffer(context[gpu_id], CL_MEM_READ_ONLY, insize,
+	    NULL, &cl_err);
+	HANDLE_CLERROR(cl_err, "Error allocating mem_in");
+	mem_salt =
+	    clCreateBuffer(context[gpu_id], CL_MEM_READ_ONLY, saltsize,
+	    NULL, &cl_err);
+	HANDLE_CLERROR(cl_err, "Error allocating mem_salt");
+	mem_out =
+	    clCreateBuffer(context[gpu_id], CL_MEM_WRITE_ONLY, outsize,
+	    NULL, &cl_err);
+	HANDLE_CLERROR(cl_err, "Error allocating mem_out");
+
+	HANDLE_CLERROR(clSetKernelArg(crypt_kernel, 0, sizeof(mem_in),
+		&mem_in), "Error while setting mem_in kernel argument");
+	HANDLE_CLERROR(clSetKernelArg(crypt_kernel, 1, sizeof(mem_out),
+		&mem_out), "Error while setting mem_out kernel argument");
+	HANDLE_CLERROR(clSetKernelArg(crypt_kernel, 2, sizeof(mem_salt),
+		&mem_salt), "Error while setting mem_salt kernel argument");
+}
+
+static void release_clobj(void)
+{
+	if (mem_in) {
+		HANDLE_CLERROR(clReleaseMemObject(mem_in), "Release mem_in");
+		HANDLE_CLERROR(clReleaseMemObject(mem_salt), "Release mem_salt");
+		HANDLE_CLERROR(clReleaseMemObject(mem_out), "Release mem_out");
+
+		MEM_FREE(inbuffer);
+		MEM_FREE(outbuffer);
+		mem_in   = NULL;
+		mem_salt = NULL;
+		mem_out  = NULL;
+	}
+}
+
+
+static void init(struct fmt_main *_self)
+{
+	self = _self;
+/*
+ Has this now become redundant? It looks like
+ autotuning does the same or similar?
+#ifdef _OPENMP
+	omp_t = omp_get_max_threads();
+	self->params.min_keys_per_crypt *= omp_t;
+	omp_t *= OMP_SCALE;
+	self->params.max_keys_per_crypt *= omp_t;
+#endif
+*/
```

Comment:
For OpenCL formats, OpenMP is only needed if we do some post-processing on host side. For this format, all `#if _OPENMP` stuff, as well as the FMT_OMP format flag, can be dropped.